File: 234-PalindromeLinkedList/234-PalindromeLinkedList.py
# Last updated: 8/13/2025, 8:20:53 PM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow = head
        fast = head

        prev = None
        slow_next = None
        while fast is not None and fast.next is not None :
            fast = fast.next.next

            slow_next = slow.next
            slow.next = prev
            prev = slow
            slow = slow_next
        
        first = None
        second = None

        if fast is not None : 
            slow = slow.next
        
        first = prev
        second = slow
        
        
        while first is not None and second is not None : 
            if first.val != second.val : 
                return False
            first = first.next
            second = second.next

        
        if first is None and second is None : 
            return True
        
        return False
        


            
            

File: 234-PalindromeLinkedList/234-PalindromeLinkedList.java
// Last updated: 8/13/2025, 8:20:57 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode first = head;
        ListNode second = head;
        Stack<ListNode> st = new Stack();

        while(first!=null){
            st.push(first);
            first = first.next;
        }

        while(st.size()>0){
            ListNode temp = st.pop();
            if(temp.val!=second.val){
                System.out.print(temp.val + " " + second.val);
                return false;
            }
            second = second.next;
        }

        return true;

    }
}